Improve performance

guyboertje opened the pull request:
- Improve the indexing creation on each (re)load.
- Move the read hot path to Java/JRuby
- Now using Sequel for remote read and local db object create only.
- Use Hikari Connection Pool.

Reviewers, please shout if you think it is weird to have part of the lookup in Ruby files and part in Java files.

Commit:
rebuild indexes after data import and update index statistics.

Commit:
Move the hot read execution to Java. Use JDBC and Hikari Connection Pool.

robbavey commented on src/main/java/org/logstash/jdbc_static/HikariCpDataSource.java:
```diff
@@ -0,0 +1,46 @@
+        config = new HikariConfig();
+        config.setJdbcUrl(url);
+        config.setPoolName("logstash.plugin.jdbc_static");
+        final int poolSize = Runtime.getRuntime().availableProcessors() + 1;
```
Should this be configurable?

guyboertje replied:
These settings are the Hikari defaults (except for the "maximumPoolSize"). I put them in here to illustrate what they are.

All my research said, only change the defaults if you benchmark the tuning process. Random changes often ended up with worse performance.

The maximum pool size comes from here.

robbavey commented on src/main/java/org/logstash/jdbc_static/HikariCpDataSource.java:
```diff
@@ -0,0 +1,46 @@
+        config.addDataSourceProperty("cachePrepStmts", "true");
+        config.addDataSourceProperty("prepStmtCacheSize", "250");
+        config.addDataSourceProperty("prepStmtCacheSqlLimit", "2048");
+
```
Configurable?

guyboertje replied:
These settings are the Hikari defaults (except for the "maximumPoolSize"). I put them in here to illustrate what they are.

All my research said, only change the defaults if you benchmark the tuning process. Random changes often ended up with worse performance.

The maximum pool size [comes from here](https://github.com/brettwooldridge/HikariCP/wiki/About-Pool-Sizing).

robbavey commented on src/test/java/org/logstash/jdbc_static/ParameterConverterTest.java:
```diff
@@ -0,0 +1,105 @@
+    public void convertString() throws SQLException {
+        ParameterConverter.convertRubyForPreparedStatement(index, ps, RUBY.newString("foo"));
+        ParameterConverter.convertRubyForPreparedStatement(index, ps, RUBY.newSymbol("foo"));
+        Mockito.verify(ps, Mockito.atLeast(2)).setString(index, "foo");
```
consider using static imports for Mockito static methods such as verify, atLeast

robbavey commented on src/test/java/org/logstash/jdbc_static/ParameterConverterTest.java:
```diff
@@ -0,0 +1,105 @@
+    @Test
+    public void convertBigDecimal() throws SQLException {
+        ParameterConverter.convertRubyForPreparedStatement(index, ps, new RubyBigDecimal(RUBY, BigDecimal.valueOf(123456789.34d)));
+        Mockito.verify(ps, Mockito.atLeast(1)).setBigDecimal(index, BigDecimal.valueOf(123456789.34d));
```
Can use atLeastOnce instead of atLeast(1)

robbavey commented on gradle.properties:
```diff
@@ -1 +1 @@
-org.gradle.daemon=false
\ No newline at end of file
+logstashCoreGemPath=/Users/guy/.rbenv/versions/jruby-9.1.13.0/lib/ruby/gems/shared/gems/logstash-core-5.6.4-java
```
As discussed on Slack, this should not be here

robbavey commented on lib/logstash/filters/jdbc/lookup.rb:
```diff
@@ -78,28 +60,58 @@ def formatted_errors
+    # any_invalid_columns? # are there any unacceptable columns in the resultset?
+    # any_invalid_parameters? # are there any parameters that did not
+    #                           A) interpolate well, B) resolve to nil, array or hash,
+    #                           C) have a type we don't coerce to a SQL type (unlikely, as we cover what Valuefier handles)
```
Maybe some documentation on the parameters/return value for this method?

guyboertje replied:
👍

robbavey commented on lib/logstash/filters/jdbc/lookup.rb:
```diff
@@ -78,28 +60,58 @@ def formatted_errors
+        tag_failure(event)
+        return false
+      end
+      true
```
It seems a little difficult to follow to have the 'success without error/warning' case dangling at the end here - I think at least a comment might be useful here. Not sure if it is a worry that if any additional failure cases arise in the `fetch_with_lock` method, they explicitly have to be handled here, otherwise they will be swallowed.

guyboertje replied:
Do you mean flipping the logic? `return true if result.all_good?` then log for the different internal errors and `tag_failure(event)` + `tag_and_set_default(event)`??

robbavey replied:
something like that, yes

guyboertje replied:
POLS, when there are invalid columns (unlikely) it means that, unless all the columns are invalid, some fields have values - so the lookup was not a complete failure. What would be the least surprising thing for a user to see?

robbavey commented on spec/filters/jdbc/lookup_spec.rb:
```diff
@@ -1,8 +1,13 @@
+require_relative "../fake_local_db"
 require "logstash/filters/jdbc/lookup"
 
+# LogStash::Logging::Logger::configure_logging("INFO")
```
Can be removed

guyboertje replied:
Sure

Commit:
add ci/run.sh for travis, make an uber jar, rework LS ruby in junit

Allow travis to build the correct gradle.properties file
Test with different versions of LS 5.4.3 to 6.X branch
Found changes in LS core that meant rework of how we load the Ruby side
in the JUnit tests.

Commit:
try fix travis

Commit:
pin rake for older LS versions

Commit:
change RubyArray#push to RubyArray#append for BWC

Commit:
prevent a triggered scheduled reload from runnning as plugin is stopping

Seen in a travis console:
`rufus-scheduler intercepted an error:`
`LogStash::Filters::Jdbc::LoaderJdbcException`
`... undefined method `[]' for nil:NilClass`
`... jdbc/read_write_database.rb:84:in `fill_local_table'`

robbavey commented on lib/logstash/filters/jdbc/read_write_database.rb:
```diff
@@ -78,14 +94,26 @@ def fill_local_table(loader)
         # There are cases where exceptions occur in unprotected ensure sections
         msg = "Exception when filling lookup db from loader #{loader.id}, query results, original exception: #{err.class}, original message: #{err.message}"
-        logger.error(msg, :backtrace => err.backtrace.take(16))
+        logger.error(msg, :backtrace => err.backtrace) #.take(16)
```
Did you mean to keep this commented out?

guyboertje replied:
Yes. I saw a stacktrace on discuss and there was not enough lines to see the logstash source.

robbavey commented on src/main/java/org/logstash/ext/RubyFetcher.java:
```diff
@@ -0,0 +1,150 @@
+        }
+
+        final IRubyObject rubyParameters = options.fastARef(ruby.newString("parameters"));
+        if (rubyParameters == null) {
```
When using a local_lookup without parameters (which worked without error on previous version of jdbc_static), this blows up with a ```java.lang.ClassCastException: org.jruby.RubyNil cannot be cast to org.jruby.RubyHash``` error. Looks like rubyParameters is set to a `RubyNil` class instead of `null`

guyboertje replied:
Hmmm? The `lookup.rb` class is supposed to validate that a parameters Hash setting was provided.
I will add the `isNil()` check though to prevent that java exception.
The `fastAref` will return null if the key is not present so, I presume, the key is present but set to nil.

robbavey reviewed (changes requested):
First set of comments from initial pass through and quick test run.

guyboertje commented on src/main/java/org/logstash/jdbc_static/HikariCpDataSource.java:
```diff
@@ -0,0 +1,46 @@
+            config.setUsername(username);
+            config.setPassword(password);
+        }
+        config.addDataSourceProperty("maximumPoolSize", String.valueOf(poolSize));
```
I can take this out it is a dup of `config.setMaximumPoolSize(poolSize)`

Commit:
updates from review

robbavey commented on lib/logstash/filters/jdbc/lookup.rb:
```diff
@@ -60,44 +69,69 @@ def formatted_errors
+        logger.warn("Parameters for the statement cannot be prepared, interpolation may have failed or the value might be nil, array or hash", "parameters" => result.invalid_parameters, "event" => event.to_hash)
+        # if this is a checking columns run and there are invalid parameters then the prepared statement is not
+        # even executed so columns cannot validated this time - need to check columns again on next event.
+        @column_validity_checked.make_false if result.checking_columns?
```
Is there a case where setting this value back to false may cause an event to 'fall through the cracks' and miss the column check? I wonder if we don't set this to true unless the columns have definitely been checked, either through a successful return or `any_invalid_columns?` being `true`. Downside - potentially some redundant logging, anything else?

guyboertje replied:
There is.

Another thread with its first event will take the `if  @column_validity_checked.true?` path and call into the Java code with column checking off. The call may complete earlier due to this - that event will not have columns checked.

Remedies:
A) Always check columns regardless.
B) Block other threads until a column check is successful.

I'm leaning towards `A`. The overhead is not so great that it warrants the deadlock risk and complexity that `B` implies.

robbavey replied:
I would tend to agree that you don't want to block until the column check is successful. The other option is to do the check until the column check is completed, instead of always, although I don't whether the trade off between code complexity and performance is worth the effort

robbavey commented on lib/logstash/filters/jdbc/lookup.rb:
```diff
@@ -60,44 +69,69 @@ def formatted_errors
-        return false
+      tag_failure(event)
+
+      if @some_valid_columns
```
Are there any tests verifying this behavior, and with multiple events making sure that the same behaviour results when column checks are done and not?

steevo590 commented:
I tested these changes in my Dev environment with production volumes, seeing a huge performance bump from previous iterations, I'm now seeing north of 4K messages a second (1 LS nodes 4 node ES cluster), where as before it barely broke 300/s

rocco8620 commented:
@guyboertje @robbavey Hello, is there a chance this is going to be merged? 

**Edit:**
I just noticed the repo has been dismissed, nevermind.
